### PR TITLE
Release via Burrito+Nix

### PR DESCRIPTION
**Pull request by rvolosatovs**
## Feature or Problem

Closes https://github.com/wasmCloud/wasmcloud-otp/issues/589 (no cross, no problem)
Closes https://github.com/wasmCloud/wasmcloud-otp/issues/572
Closes https://github.com/wasmCloud/wasmcloud-otp/issues/570

- No QEMU/Docker emulation used to build, only cross or native compilation
- Bakeware -> Burrito
- Builds for aarch64 and x86_64 glibc *and* musl toolchains
- Build caching
- Fully reproducible builds for everything except:
	- Mac NIFs
	- Windows NIF
- Link against GLIBC 2.29, which is compatible with Ubuntu 20.04 and later

### Dependencies:
- MacOS: none
- Ubuntu:
	- `ca-certificates` - self-explanatory
- Alpine:
	- `libgcc` - that's actually a NIF dependency, which is added by the Rust compiler https://github.com/rust-lang/rust/issues/82521
- Windows: none

After setting the required env variables the release can be reproduced via `release` within a nix shell (`nix develop` from anywhere within the repo or `nix develop github:wasmcloud/wasmcloud-otp` once the PR is merged)
There's also a `make release` target added in the root, which initializes a Nix shell and calls `release`

Note, that you have to first:
- build a Mac NIF for both architectures on a Mac and set appropriate env vars
- build a Windows NIF on a Windows box and set appropriate env vars
- set `SECRET_KEY_BASE` in your environment

Note, that NIFs can also be just fetched from one of CI runs

e.g. I have this in my `.envrc` (you may want to look into https://direnv.net/)
```
export SECRET_KEY_BASE="3ImiTAMO0TTD7wrACHrCA+ggkzpw6zGWvE3gtQwlXE6vmnDT9yGP5/WKpLWEJ8fF"
export HOSTCORE_WASMCLOUD_NATIVE_AARCH64_APPLE_DARWIN="path://$PWD/_nix/aarch64"
export HOSTCORE_WASMCLOUD_NATIVE_X86_64_APPLE_DARWIN="path://$PWD/_nix/x86_64"
export HOSTCORE_WASMCLOUD_NATIVE_X86_64_PC_WINDOWS_MSVC="path://$PWD/_nix/windows"
```

(you have to first fetch and unzip the libraries just like CI does, see workflow and error messages, if any)

## Testing

Built on platform(s)
- [x] x86_64-linux
- [x] aarch64-linux
- [x] x86_64-darwin
- [ ] aarch64-darwin
- [ ] x86_64-windows

Tested on platform(s)
- [x] x86_64-linux
- [ ] aarch64-linux
- [x] x86_64-darwin
- [x] aarch64-darwin
- [x] x86_64-windows

### CODE REVIEW

**Comment by ricochet** [comment]
The x86 darwin error is interesting. Failed to load workspace members `dependency.wasi-cap-std-sync` was not found in `workspace.dependencies`

https://github.com/wasmCloud/wasmcloud-otp/actions/runs/4836354104/jobs/8619473766?pr=616#step:4:2319

**Comment by rvolosatovs** [comment]
> The x86 darwin error is interesting. Failed to load workspace members `dependency.wasi-cap-std-sync` was not found in `workspace.dependencies`
> 
> https://github.com/wasmCloud/wasmcloud-otp/actions/runs/4836354104/jobs/8619473766?pr=616#step:4:2319

Yes, but it actually does not seem to influence the build, just a yet another preview2-prototyping oddity

**Comment by rvolosatovs** [comment]
See also https://github.com/wasmCloud/wasmcloud-otp/issues/608#issuecomment-1538379649

**Comment by brooksmtownsend** [comment]
> aarch64 is only compatible with Ubuntu 22.04 and later for now

@rvolosatovs is this required? I believe this will break the raspberry pi's running on debian:bullseye, which is the latest raspbian

Great callout for `ca-certificates`, which I'm sure is present in many Linux installations but we can explicitly call it out. I'll say that this worked for me on my Mac, and in a docker container, a previous test of Windows also worked great but i can give it another go as well.

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -2,380 +2,413 @@ name: WasmCloud Release
...
+        name: ${{ matrix.package }}_x86_64_windows
+        path: ./result/bin/${{ matrix.package }}_x86_64_windows.exe
+        if-no-files-found: error
+
+  test-darwin:
+    runs-on: macos-latest
+    needs:
+    - burrito-darwin
+    steps:
+    - run: curl -sL 'https://github.com/nats-io/nats-server/releases/download/v2.9.16/nats-server-v2.9.16-darwin-amd64.tar.gz' | tar xzv --strip-components 1
```
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply .github/workflows/release.yml:197]: https://github.com/wasmCloud/common-actions/issues/10

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -2,380 +2,413 @@ name: WasmCloud Release
...
+        timeout /t 30
+        Taskkill /im wasmcloud_host_x86_64_windows.exe /f
+
+  release:
+    # TODO: Build on PRs and pushes to main (but do not push)
+    # This is currently blocked on the `app-version` computation, which only supports semver tags
+    # E.g.:
+    # ```
+    # Error: buildx failed with: ERROR: invalid tag "wasmcloud.azurecr.io/wasmcloud_host:616/merge": invalid reference format
+    # ```
```
Comment (brooksmtownsend) [inline .github/workflows/release.yml:267]: Leaving comment for later, if we want to run canary releases or something this sounds good. We can create an issue

**Review by brooksmtownsend — COMMENTED** [review]
This is all fantastic work 🎉 As mentioned in the other comment I verified this works for mac/linux aarch64, will test x86 windows again after leaving this review.

I left a few nit comments, the one that we should probably address is using the `-latest` runners or pinned versions, I've opted for the pinned versions just to avoid our pipelines breaking out from under us when runners update, but it's not a hard-held stance

**Review by brooksmtownsend — COMMENTED** [review]
This is all fantastic work 🎉 As mentioned in the other comment I verified this works for mac/linux aarch64, will test x86 windows again after leaving this review.

I left a few nit comments, the one that we should probably address is using the `-latest` runners or pinned versions, I've opted for the pinned versions just to avoid our pipelines breaking out from under us when runners update, but it's not a hard-held stance.

Edit: Windows works 🎉

File: wasmcloud_host/Dockerfile [hunk 3]
```diff
@@ -1,146 +1,24 @@
...
 ARG WASMCLOUD_CTL_HOST=nats
-ENV WASMCLOUD_RPC_HOST=${WASMCLOUD_RPC_HOST} \
+ENV LANG="C.UTF-8" \
+    LC_TYPE="C.UTF-8" \
+    WASMCLOUD_RPC_HOST=${WASMCLOUD_RPC_HOST} \
     WASMCLOUD_CTL_HOST=${WASMCLOUD_CTL_HOST} \
     WASMCLOUD_PROV_RPC_HOST=${WASMCLOUD_PROV_RPC_HOST}
 
-CMD ["/opt/app/bin/wasmcloud_host", "start"]
+CMD ["/usr/local/bin/wasmcloud_host"]
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply wasmcloud_host/Dockerfile:24]: That's already done https://github.com/wasmCloud/wasmcloud-otp/pull/616/files#diff-0d14355dd2b18d5b715d38df326ec53a9999a82be0ed5708597189b9b809394fR11

File: host_core/Dockerfile [hunk 4]
```diff
@@ -1,61 +1,24 @@
...
-ENV WASMCLOUD_RPC_HOST=${WASMCLOUD_RPC_HOST} \
+ENV LANG="C.UTF-8" \
+    LC_TYPE="C.UTF-8" \
+    WASMCLOUD_RPC_HOST=${WASMCLOUD_RPC_HOST} \
     WASMCLOUD_CTL_HOST=${WASMCLOUD_CTL_HOST} \
     WASMCLOUD_PROV_RPC_HOST=${WASMCLOUD_PROV_RPC_HOST}
 
-CMD ["/opt/app/bin/host_core", "start"]
\ No newline at end of file
+CMD ["/usr/local/bin/host_core"]
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply host_core/Dockerfile]: That's already done

**Review by brooksmtownsend — APPROVED** [review]
LGTM once tests pass, works for me on aarch64 linux, mac, and x86 windows

**Review by brooksmtownsend — APPROVED** [review]
LGTM once tests pass, works for me on aarch64 linux, mac, and x86 windows

**Comment by rvolosatovs** [comment]
> LGTM once tests pass, works for me on aarch64 linux, mac, and x86 windows

Could you try to restart them? (Potentially a few times) Since I have not changed the implementation, it appears as a flake to me.

Otherwise, could you tell me what is there to be fixed? There are way too many logs and it's impossible to figure out what's wrong for someone unfamiliar with the project (like myself)

**Review by brooksmtownsend — APPROVED** [review]

**Pull request merged**